### PR TITLE
Neural network potential with the n2p2 package

### LINKED ISSUE #7: Neural network potential
Add initial implementation of NN force kernel

### PR DESCRIPTION
**Pull request by saakethdesai**
Closes #7

### CODE REVIEW

**Comment by junghans** [comment]
I think it would be better to use CMake's External Project feature instead of copying the source into our repo.

Something like:
```cmake
include(ExternalProject)
ExternalProject_Add(p2n2_build
      URL https://github.com/CompPhysVienna/n2p2/archive/v1.0.0.tar.gz
      CONFIGURE_COMMAND "" 
      SOURCE_SUBDIR src
      BUILD_COMMAND make libnnp 
      BUILD_IN_SOURCE 1
      INSTALL_COMMAND ""
      )
ExternalProject_get_property(p2np_build SOURCE_DIR)
set(P2N2_LIBRARIES ${SOURCE_DIR}/src/libnnp.a)
set(P2N2_INCLUDE_DIRS ${SOURCE_DIR}/src)
add_dependencies(CabanaMD p2np_build)
```

**Comment by streeve** [comment]
> I think it would be better to use CMake's External Project feature instead of copying the source into our repo.

That's a good point. @saakethdesai , can you switch to an external project and copy in only the files you need to modify (renamed appending with Cabana)?

**Comment by junghans** [comment]
Another option is:
```cmake
include(FetchContent)
FetchContent_Declare(n2p2 GIT_REPOSITORY https://github.com/CompPhysVienna/n2p2.git GIT_TAG 1.0.0)
```

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 *~
-build
```
Comment (streeve) [inline .gitignore]: Revert

File: src/cabanamd.cpp [hunk 2]
```diff
@@ -80,7 +80,7 @@ void CabanaMD::init(int argc, char* argv[]) {
   input->read_file();
   printf("Read input file\n");
   T_X_FLOAT neigh_cutoff = input->force_cutoff + input->neighbor_skin;
-
+
```
Comment (streeve) [inline src/cabanamd.cpp]: Spacing

File: src/cabanamd.cpp [hunk 3]
```diff
@@ -131,7 +131,7 @@ void CabanaMD::init(int argc, char* argv[]) {
   Cabana::deep_copy(f, 0.0);
   force->compute(system);
 
-  if(input->comm_newton) {
+  if(input->comm_newton || input->force_type == 2) { //update force if nnp pair style
```
Comment (streeve) [inline src/cabanamd.cpp]: Use enum

File: src/cabanamd.cpp [hunk 4]
```diff
@@ -229,7 +229,7 @@ void CabanaMD::run(int nsteps) {
     // This is where Bonds, Angles and KSpace should go eventually 
     
     // Reverse Communicate Force Update on Halo
-    if(input->comm_newton) {
+    if(input->comm_newton or input->force_type == 2) { //update force if nnp pair style
```
Comment (streeve) [inline src/cabanamd.cpp]: Also use enum

File: src/cabanamd.cpp [hunk 5]
```diff
@@ -267,7 +267,16 @@ void CabanaMD::run(int nsteps) {
 
     other_time += other_timer.seconds();
   }
-
+    /*auto f = Cabana::slice<Forces>(system->xvf);
```
Comment (streeve) [inline src/cabanamd.cpp]: Remove

File: src/force_types/force_nnp_cabana_neigh.h [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    if( (strcmp(argv[i+1], "NEIGH_CSR") == 0) )
+      neighbor_type = NEIGH_CSR;
+  }
+#endif
+#ifdef FORCE_MODULES_INSTANTIATION
+    else if (input->force_type == FORCE_NNP) {
+      bool half_neigh = input->force_iteration_type == FORCE_ITER_NEIGH_HALF;
+      if (input->neighbor_type == NEIGH_2D) {
+        if (half_neigh)
+          force = NULL;
```
Comment (streeve) [inline src/force_types/force_nnp_cabana_neigh.h]: Give some feedback to the user that this is not an option (for all 3)?

File: src/read_data.h [hunk 7]
```diff
@@ -163,8 +162,9 @@ void read_lammps_atoms(ifstream &file, System* s)
          (x_tmp <  s->sub_domain_hi_x) &&
          (y_tmp <  s->sub_domain_hi_y) &&
          (z_tmp <  s->sub_domain_hi_z) ) { 
-        id(n) = id_tmp; type(n) = type_tmp; x(n,0) = x_tmp; x(n,1) = y_tmp; x(n,2) = z_tmp;
+        id(n) = id_tmp; type(n) = type_tmp-1; x(n,0) = x_tmp; x(n,1) = y_tmp; x(n,2) = z_tmp;
```
Comment (streeve) [inline src/read_data.h]: Maybe separate into a "fix read_data" PR

File: src/system.h [hunk 8]
```diff
@@ -68,6 +68,10 @@ class System {
   // Per Type Property
   t_mass mass;
 
+  //NNP conversion factors (TODO: remove hardcoding)
+  double cflength = 1.889726;
```
Comment (streeve) [inline src/system.h]: Move to input / force_nnp

File: src/types.h [hunk 9]
```diff
@@ -109,6 +110,8 @@ using t_tuple = Cabana::MemberTypes<T_FLOAT[3], T_FLOAT[3], T_FLOAT[3],
                                     T_INT, T_INT, T_FLOAT>;
 enum TypeNames { Positions = 0, Velocities = 1, Forces = 2,
                  Types = 3, IDs = 4, Charges = 5 };
+enum NNPNames { G = 0, dEdG = 1, energy = 2};
```
Comment (streeve) [inline src/types.h]: This should really be in force_nnp or mode, but I'm fine with it here for now

File: src/property_pote.cpp [hunk 10]
```diff
@@ -49,6 +49,7 @@
 
 #include <cabanamd.h>
 #include <property_pote.h>
+#include <iostream>
```
Comment (streeve) [inline src/property_pote.cpp]: Is this necessary?

File: src/libnnp/Mode.cpp [hunk 11]
```diff
@@ -0,0 +1,1316 @@
...
+    Cabana::deep_copy(G,0.0); 
+
+    // Calculate symmetry functions (and derivatives).
+    //calculateSFGroups(s, nnp_data, SF, SFscaling, SFGmemberlist, attype, neigh_list, i, countergtotal);
+    //Eval functor;
+    //Kokkos::parallel_for (s->N_local, functor);
+    //Kokkos::fence();
+    timer.reset();
+    /*printf("Positions : \n");
+    printf("%d %f %f %f\n", id(0), x(0,0), x(0,1), x(0,2));
```
Comment (streeve) [inline src/libnnp/Mode.cpp]: Remove prints

File: src/force_types/force_nnp_cabana_neigh.cpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+void ForceNNP::create_neigh_list(System* system) {
+  N_local = system->N_local;
+  double grid_min[3] = {system->sub_domain_lo_x - system->sub_domain_x, 
+    system->sub_domain_lo_y - system->sub_domain_y, system->sub_domain_lo_z - system->sub_domain_z};
+  double grid_max[3] = {system->sub_domain_hi_x + system->sub_domain_x, 
+    system->sub_domain_hi_y + system->sub_domain_y, system->sub_domain_hi_z + system->sub_domain_z};
+
+  auto x = Cabana::slice<Positions>(system->xvf);
+  auto id = Cabana::slice<IDs>(system->xvf);
```
Comment (streeve) [inline src/force_types/force_nnp_cabana_neigh.cpp]: Unused slice

File: src/libnnp/Mode.cpp [hunk 13]
```diff
@@ -0,0 +1,1316 @@
...
+
+
+
+
+
+
+void Mode::calculateSymmetryFunctionGroups(System *s, AoSoA_NNP nnp_data, t_verletlist_full_2D neigh_list,
+    t_mass numSymmetryFunctionsPerElement) 
+{
+    auto id = Cabana::slice<IDs>(s->xvf);
```
Comment (streeve) [inline src/libnnp/Mode.cpp]: Unused slice

File: src/libnnp/Mode.cpp [hunk 14]
```diff
@@ -0,0 +1,1316 @@
...
+    printf("%d %f %f %f\n", id(2), G(2,0), G(2,1), G(2,2));    
+    printf("%d %f %f %f\n", id(3), G(3,0), G(3,1), G(3,2));    
+    printf("%d %f %f %f\n", id(4), G(4,0), G(4,1), G(4,2));    
+    printf("%d %f %f %f\n", id(5), G(5,0), G(5,1), G(5,2));*/    
+} 
+
+void Mode::calculateAtomicNeuralNetworks(System* s, AoSoA_NNP nnp_data, t_mass numSymmetryFunctionsPerElement)
+{
+    //Calculate Atomic Neural Networks 
+    auto id = Cabana::slice<IDs>(s->xvf);
```
Comment (streeve) [inline src/libnnp/Mode.cpp]: Unused slice

**Review by streeve — CHANGES_REQUESTED** [review]
The major thing to update before merging is removing the files unmodified from libnnp, linking them in from a separate build (just like we do with Kokkos/Cabana), overloading just the functions that you've replaced, and removing the rest. Eventually we can do a submodule, but let's focus on removing overlap first

File: src/libnnp/Mode.cpp [hunk 15]
```diff
@@ -0,0 +1,1316 @@
...
+
+    NN = d_t_NN("Mode::NN",s->N,numLayers,maxNeurons);
+    dfdx = d_t_NN("Mode::dfdx",s->N,numLayers,maxNeurons);
+    inner = d_t_NN("Mode::inner",s->N,numHiddenLayers,maxNeurons);
+    outer = d_t_NN("Mode::inner",s->N,numHiddenLayers,maxNeurons);
+    
+    timer.reset();
+    Kokkos::parallel_for ("Mode::calculateAtomicNeuralNetworks", s->N_local, KOKKOS_LAMBDA (const int atomindex)
+    {
+        for (int i = 0; i < numLayers; ++i)
```
Comment (streeve) [inline src/libnnp/Mode.cpp]: Use deep_copy for each view

File: src/libnnp/Mode.cpp [hunk 16]
```diff
@@ -0,0 +1,1316 @@
...
+            else layer_lminusone = numNeuronsPerLayer[l-1];
+            double dtmp;
+            for (int i = 0; i < numNeuronsPerLayer[l]; i++)
+            {
+                dtmp = 0.0;
+                for (int j = 0; j < layer_lminusone; j++)
+                    dtmp += weights(attype,l-1,i,j) * NN(atomindex,l-1,j);
+                dtmp += bias(attype,l-1,i);
+                //if (normalizeNeurons) dtmp /= numNeuronsPerLayer[l-1];
+                if (AF[l] == 0)
```
Comment (streeve) [inline src/libnnp/Mode.cpp]: Maybe View instead?

**Review by streeve — CHANGES_REQUESTED** [review]
Couple extra notes from our discussion

**Comment by streeve** [comment]
@saakethdesai I think this is ready, let me know what you think 

As a note, I left the on-node version as a separate branch (intended to keep for comparison with the paper)

**Review by streeve — APPROVED** [review]

**Pull request merged**